From a848f2de27d48d7b90c15c325fcfe19a1eaec339 Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Mon, 3 Sep 2012 17:39:20 +0800
Subject: [PATCH 2/3] update .gitignore

---
 .gitignore | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/.gitignore b/.gitignore
index 5041ddf..b98059e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -191,3 +191,8 @@ stamp-h1
 !/po/*.po
 !/po/POTFILES.in
 !/po/libvirt.pot
+
+log_*
+patch*
+build.sh
+
-- 
1.7.12

