From 7ebe05149e59b345d5921e41ff9cec0357189d06 Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Mon, 3 Sep 2012 21:15:36 +0800
Subject: [PATCH 3/3] add ATTRIBUTE_UNUSED for fixing compiling warning

---
 src/libxl/libxl_driver.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/libxl/libxl_driver.c b/src/libxl/libxl_driver.c
index 483e2cb..ccc7051 100644
--- a/src/libxl/libxl_driver.c
+++ b/src/libxl/libxl_driver.c
@@ -406,7 +406,7 @@ libxlDomainObjEndAsyncJob(libxlDriverPrivatePtr driver, virDomainObjPtr obj)
     return virObjectUnref(obj);
 }
 
-static int
+static int ATTRIBUTE_UNUSED
 libxlMigrationJobStart(libxlDriverPrivatePtr driver,
                        virDomainObjPtr vm,
                        enum libxlDomainAsyncJob job)
@@ -424,7 +424,7 @@ libxlMigrationJobStart(libxlDriverPrivatePtr driver,
     return 0;
 }
 
-static bool
+static bool ATTRIBUTE_UNUSED
 libxlMigrationJobFinish(libxlDriverPrivatePtr driver, virDomainObjPtr vm)
 {
     return libxlDomainObjEndAsyncJob(driver, vm);
-- 
1.7.12

