From d9cba0024416e6a0635081744512bda19b17c7ba Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Tue, 4 Sep 2012 13:42:46 +0800
Subject: [PATCH 3/4] update .gitignore according to syntax-check

---
 .gitignore | 1 -
 1 file changed, 1 deletion(-)

diff --git a/.gitignore b/.gitignore
index b98059e..0475b58 100644
--- a/.gitignore
+++ b/.gitignore
@@ -195,4 +195,3 @@ stamp-h1
 log_*
 patch*
 build.sh
-
-- 
1.7.12

