From 423319598131d20f212678598675250d75f1a3c9 Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Mon, 3 Sep 2012 20:43:32 +0800
Subject: [PATCH 04/17] fix compile warnings

---
 src/libxl/libxl_driver.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/src/libxl/libxl_driver.c b/src/libxl/libxl_driver.c
index 18d6004..bff0d4a 100644
--- a/src/libxl/libxl_driver.c
+++ b/src/libxl/libxl_driver.c
@@ -354,8 +354,8 @@ libxlDomainObjBeginJobWithDriver(libxlDriverPrivatePtr driver,
     return libxlDomainObjBeginJobInternal(driver, true, obj, job,
                                          LIBXL_ASYNC_JOB_NONE);
 }
-/*
-static int
+
+static int ATTRIBUTE_UNUSED
 libxlDomainObjBeginAsyncJobWithDriver(libxlDriverPrivatePtr driver,
                                          virDomainObjPtr obj,
                                          enum libxlDomainAsyncJob asyncJob)
@@ -363,7 +363,6 @@ libxlDomainObjBeginAsyncJobWithDriver(libxlDriverPrivatePtr driver,
     return libxlDomainObjBeginJobInternal(driver, true, obj, LIBXL_JOB_ASYNC,
                                          asyncJob);
 }
-*/
 
 /*
  * obj must be locked before calling, libxlDriverPrivatePtr does not matter
@@ -407,7 +406,7 @@ libxlDomainObjEndAsyncJob(libxlDriverPrivatePtr driver, virDomainObjPtr obj)
     return virObjectUnref(obj);
 }
 
-static int
+static int ATTRIBUTE_UNUSED
 libxlMigrationJobStart(libxlDriverPrivatePtr driver,
                        virDomainObjPtr vm,
                        enum libxlDomainAsyncJob job)
@@ -425,7 +424,7 @@ libxlMigrationJobStart(libxlDriverPrivatePtr driver,
     return 0;
 }
 
-static bool
+static bool ATTRIBUTE_UNUSED
 libxlMigrationJobFinish(libxlDriverPrivatePtr driver, virDomainObjPtr vm)
 {
     return libxlDomainObjEndAsyncJob(driver, vm);
-- 
1.7.12

