From 55db6e43102b53694ff889b6a38928762321bea1 Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Mon, 3 Sep 2012 21:10:31 +0800
Subject: [PATCH 05/17] fix tailing blank

---
 src/libxl/libxl_driver.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/libxl/libxl_driver.c b/src/libxl/libxl_driver.c
index bff0d4a..334e0fa 100644
--- a/src/libxl/libxl_driver.c
+++ b/src/libxl/libxl_driver.c
@@ -1934,7 +1934,7 @@ libxlDomainDestroyFlags(virDomainPtr dom,
     }
 
     if (!vm->persistent) {
-        if (libxlDomainObjEndJob(driver, vm)) 
+        if (libxlDomainObjEndJob(driver, vm))
             virDomainRemoveInactive(&driver->domains, vm);
         vm = NULL;
     }
@@ -2274,7 +2274,7 @@ libxlDoDomainSave(libxlDriverPrivatePtr driver, virDomainObjPtr vm,
     ret = libxl_domain_suspend(&priv->ctx, NULL, vm->def->id, fd);
     libxlDriverLock(driver);
     virDomainObjLock(vm);
- 
+
     if (ret != 0) {
         virReportError(VIR_ERR_INTERNAL_ERROR,
                        _("Failed to save domain '%d' with libxenlight"),
@@ -2397,7 +2397,7 @@ libxlDomainRestoreFlags(virConnectPtr conn, const char *from,
 
     if ((ret = libxlVmStart(driver, vm, false, fd)) < 0 &&
         !vm->persistent) {
-        if (libxlDomainObjEndAsyncJob(driver, vm)) 
+        if (libxlDomainObjEndAsyncJob(driver, vm))
             virDomainRemoveInactive(&driver->domains, vm);
         vm = NULL;
     }
@@ -3167,7 +3167,7 @@ libxlDomainCreateWithFlags(virDomainPtr dom,
 endjob:
     if (!libxlDomainObjEndJob(driver, vm))
         vm = NULL;
- 
+
 cleanup:
     if (vm)
         virDomainObjUnlock(vm);
-- 
1.7.12

