From 43e55aa7ed7f928e142d0122eed1ba3663cd66b2 Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Mon, 10 Sep 2012 15:18:24 +0800
Subject: [PATCH 15/17] add debug info for begin job and end job

---
 src/libxl/libxl_driver.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/libxl/libxl_driver.c b/src/libxl/libxl_driver.c
index 260d2a8..df8330f 100644
--- a/src/libxl/libxl_driver.c
+++ b/src/libxl/libxl_driver.c
@@ -229,6 +229,8 @@ libxlDomainObjBeginJobInternal(libxlDriverPrivatePtr driver,
     unsigned long long now;
     unsigned long long then;
 
+    VIR_INFO("enter");
+
     if (virTimeMillisNow(&now) < 0)
         return -1;
     then = now + LIBXL_JOB_WAIT_TIME;
@@ -310,6 +312,7 @@ error:
         virDomainObjLock(obj);
     }
     virObjectUnref(obj);
+    VIR_INFO("exit");
     return -1;
 }
 
@@ -389,6 +392,7 @@ libxlDomainObjEndJob(libxlDriverPrivatePtr driver, virDomainObjPtr obj)
     libxlDomainObjPrivatePtr priv = obj->privateData;
     enum libxlDomainJob job = priv->job.active;
 
+    VIR_INFO("enter");
     VIR_DEBUG("Stopping job: %s (async=%s)",
               libxlDomainJobTypeToString(job),
               libxlDomainAsyncJobTypeToString(priv->job.asyncJob));
@@ -398,6 +402,7 @@ libxlDomainObjEndJob(libxlDriverPrivatePtr driver, virDomainObjPtr obj)
         libxlDomainObjSaveJob(driver, obj);
     virCondSignal(&priv->job.cond);
 
+    VIR_INFO("exit");
     return virObjectUnref(obj);
 }
 
-- 
1.7.12

