From 12dbd725ef60c078d2bc2ab8ad92c4c3b87de252 Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Mon, 10 Sep 2012 17:21:23 +0800
Subject: [PATCH 16/17] disable libvirtd log in syslog

---
 daemon/libvirtd.conf | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/daemon/libvirtd.conf b/daemon/libvirtd.conf
index 7330a34..c105605 100644
--- a/daemon/libvirtd.conf
+++ b/daemon/libvirtd.conf
@@ -327,7 +327,7 @@ log_level = 2
 #
 # Multiple output can be defined, they just need to be separated by spaces.
 # e.g. to log all warnings and errors to syslog under the libvirtd ident:
-log_outputs="2:syslog:libvirtd"
+#log_outputs="2:syslog:libvirtd"
 #
 
 # Log debug buffer size: default 64
-- 
1.7.12

