From ada631efb97539f21fc121286770243123cf487a Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Mon, 10 Sep 2012 17:26:57 +0800
Subject: [PATCH 17/17] update daemon/libvirtd.conf

---
 daemon/libvirtd.conf | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/daemon/libvirtd.conf b/daemon/libvirtd.conf
index 3d296cd..c105605 100644
--- a/daemon/libvirtd.conf
+++ b/daemon/libvirtd.conf
@@ -164,15 +164,15 @@
 
 # Override the default server key file path
 #
-#key_file = "/etc/pki/libvirt/private/serverkey.pem"
+key_file = "/etc/pki/libvirt/private/serverkey.pem"
 
 # Override the default server certificate file path
 #
-#cert_file = "/etc/pki/libvirt/servercert.pem"
+cert_file = "/etc/pki/libvirt/servercert.pem"
 
 # Override the default CA certificate path
 #
-#ca_file = "/etc/pki/CA/cacert.pem"
+ca_file = "/etc/pki/CA/cacert.pem"
 
 # Specify a certificate revocation list.
 #
@@ -286,7 +286,7 @@
 
 # Logging level: 4 errors, 3 warnings, 2 information, 1 debug
 # basically 1 will log everything possible
-#log_level = 3
+log_level = 2
 
 # Logging filters:
 # A filter allows to select a different logging level for a given category
@@ -327,7 +327,7 @@
 #
 # Multiple output can be defined, they just need to be separated by spaces.
 # e.g. to log all warnings and errors to syslog under the libvirtd ident:
-#log_outputs="3:syslog:libvirtd"
+#log_outputs="2:syslog:libvirtd"
 #
 
 # Log debug buffer size: default 64
-- 
1.7.12

