--- /etc/bash.bashrc	2016-10-29 00:26:38.000000000 +0800
+++ /etc/bash.bashrc.new	2016-11-29 10:57:09.717633924 +0800
@@ -255,6 +255,7 @@
 	test -s /etc/profile.d/alias.ash && . /etc/profile.d/alias.ash
     else
 	test -s /etc/profile.d/alias.bash && . /etc/profile.d/alias.bash
+	test -s /etc/profile.d/alias.bash.bamvor && . /etc/profile.d/alias.bash.bamvor
 	test -s $HOME/.alias && . $HOME/.alias
     fi
 
