From c89de311a547398ef3cfe54f9fe284b15cfac370 Mon Sep 17 00:00:00 2001
From: Bamvor Jian Zhang <bjzhang@suse.com>
Date: Tue, 3 Dec 2013 11:10:42 +0800
Subject: [PATCH 3/6] fix compile error

---
 src/libxl/libxl_driver.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/libxl/libxl_driver.c b/src/libxl/libxl_driver.c
index c0bb8cc..7ca3ce4 100644
--- a/src/libxl/libxl_driver.c
+++ b/src/libxl/libxl_driver.c
@@ -66,10 +66,12 @@
 /* Number of Xen scheduler parameters */
 #define XEN_SCHED_CREDIT_NPARAM   2
 
-#define LIBXL_AO_SIGHANDLER
+//#define LIBXL_AO_SIGHANDLER
 
 static libxlDriverPrivatePtr libxl_driver = NULL;
+#ifdef LIBXL_AO_SIGHANDLER
 static struct sigaction sigchld_saved_action;
+#endif
 
 /* Function declarations */
 static int
-- 
1.8.1.4

